Update procedure Registering a Host to RH Satellite (#780)
theforeman/foreman-documentation: guides/doc-Managing_Hosts/topics/proc_registering-a-host.adoc

For Red Hat family distributions, enter the URL of the repository, for example `\http://rpm.example.com/`.
ifndef::satellite[]
For Debian OS families, enter the whole line of list file content, for example `deb \http://deb.example.com/ buster 1.0`.
endif::[]
. Optional: In the *Repository GPG key URL* field, specify the public key to verify the signatures of GPG-signed packages.
It needs to be specified in the ASCII form with the GPG public key header.

ifdef::satellite,orcharhino[]
. In the *Activation Key(s)* field, enter one or more activation keys to assign to hosts.
. Optional: Select the *Lifecycle environment*.
. Optional: Select the *Ignore errors* option if you want to ignore subscription manager errors.
. Optional: Select the *Force* option if you want to remove any `katello-ca-consumer` rpms before registration and run subscription-manager with the `--force` argument.
endif::[]

ifdef::foreman-el,foreman-deb,katello[]
. Optional: This step is for the Katello users only.
If you register RHEL or CentOS hosts, in the *Activation Key(s)* field, enter one or more activation keys to assign to registered hosts.
endif::[]

> . Click the *Generate* button.

Author: @stejskalleos What does it mean practically? That I don't make this change? Or something else?

> . Click the *Generate* button.

Reviewer: Changes in https://github.com/theforeman/foreman-documentation/pull/745  will probably conflict with the changes in this PR, so would be good to wait until the [#745](https://github.com/theforeman/foreman-documentation/pull/745) is merged

> . Click the *Generate* button.

Author: Understood and I see that #745 has been merged, so I'm proceeding with a rebase.

.Procedure
. Navigate to *Hosts* > *Register Host*.
. Optional: Select a different *Organization*.
. Optional: Select a different *Location*.
. Optional: From the *Host Group* list, select the host group to associate the hosts with.

> Fields that inherit value from *Host group*: Operating system, Activation Key(s) and Lifecycle environment.

Reviewer 2: IBM Style guide says `no` to adding a `(s)`. See "Parenthese to form plural" in chapter 2.

Author: Actually, the field in the WebUI is labelled with that precise string, are we sure we want to introduce an inconsistency with the WebUI for the sake of the SG?

Author: Perhaps it would be better to markup those field labels with emphasis instead.

Author: Oh, sorry, my mistake. I've checked the WebUI and it follows the SG.

. Optional: From the *{SmartProxy}* list, select the {SmartProxy} to register hosts through.

> ifdef::satellite[]

Reviewer 2: Is this only relevant for Satellite, but not Foreman?

You must select the internal {SmartProxy} if you do not want to use an external {SmartProxy}.
endif::[]
. Optional: From the *Operating system* list, select the operating system of hosts that you want to register.
ifdef::satellite[]

> Specifying an operating system is required when you register machines without subscription-manager, such as Debian or Ubuntu.

Reviewer 2: ```suggestion
Specifying an operating system is required when you register machines without `subscription-manager`, such as Debian or Ubuntu.
```

Are you sure you want to mention Debian and Ubuntu in Satellite docs? Did you mean to go with `ifndef`?

Author: Good point, thanks! ifndef is probably a way to go.

endif::[]
. Optional: Select the *Insecure* option, if you want to make the first call insecure.
During this first call, hosts download the CA file from {Project}.
Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+
If an attacker, located in the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JSON Web Tokens (JWT).
Therefore, if you have chosen to deploy SSH keys during registration, the attacker will be able to access the host using the SSH key.
+
The following is an example of the `curl` command with the `--insecure` option:
+
[options="nowrap", subs="+quotes,attributes"]
----

> curl -sS --insecure https://{foreman-example-com}/register...

Reviewer 3: I guess this needs some input from @stejskalleos, my wild guess is `/unattended/register` but not sure about the rest.

Also note we should not use `-k` or `--insecure` in our docs, we should always ask our users to install CA certificate. I think this could be modularized too:

```
guides/common/modules/proc_configuring-satellite-for-outgoing-emails.adoc
77:# update-ca-trust enable
78:# update-ca-trust

guides/doc-Administering_Red_Hat_Satellite/topics/Configuring_Identity_Management.adoc
226:# update-ca-trust enable
227:# update-ca-trust

guides/doc-Managing_Hosts/topics/proc_registering-a-host.adoc
112:# update-ca-trust enable
113:# update-ca-trust

guides/doc-Provisioning_Guide/topics/Infoblox_Integration.adoc
40:# update-ca-trust enable
43:# update-ca-trust extract

guides/doc-Provisioning_Guide/topics/Virt-VMWare.adoc
374:# update-ca-trust enable
391:# update-ca-trust extract
```

Author: @stejskalleos Can you please clarify what the proper example URL should look like?

Reviewer: the URL is `https://foreman.example.com/register`

----
+
If you do not want to call the `curl` command with the `--insecure` option, you can manually copy and install the CA file on each host.
+
To do this, find where {Project} stores the CA file by navigating to *Administer* > *Settings* > *Authentication* and locating the value of the *SSL CA file* setting.
+
Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on hosts and enter the following commands:
+
[options="nowrap", subs="+quotes,attributes"]
----
# update-ca-trust enable
# update-ca-trust

> ----

Reviewer 3: Allright, I was too fast. Well, in that case make this the default and move it above. Avoid mentioning `--insecure` please, we do not want users to do insecure setups.

Author: @lzap In that case, why do we have it in the WebUI in the first place?

Author: I will do the changes you requested, I'm just saying it might be something to consider in UX.

Reviewer: > why do we have the Insecure option in the WebUI in the first place?

Good question. We should users encourage to not use the `insecure`, because it is insecure (duh!).

With `--insecure` flag, they just need to run registration command and host is registered.  It is simple and easy.
Without `insecure`, users have to prior the registration connect to the hosts, deploy the certificates and run `update-ca-trust`. After that they can run the registration command and register the host.

They have to choose between being comfortable and being secure.

. Select the *Advanced* tab.
. From the *Setup REX* list, select whether you want to deploy {Project} SSH keys to hosts or not.
+

> If set to `Yes`, SSH keys will be installed on the registered host.

Reviewer 2: ```suggestion
If set to `Yes`, public SSH keys will be installed on the registered host.
```

The inherited value is based on the `host_registration_remote_execution` parameter.

> It can be inherited e.g. from host group, operating system, organization.

Reviewer 2: ```suggestion
It can be inherited, for example from a host group, an operating system, or an organization.
```

When overridden, the selected value will be stored on host parameter level.

. From the *Setup Insights* list, select whether you want to install `insights-client` and register the hosts to Insights.
+

> The Insights tool is available for {RHEL} only. It has no effect on other operating systems.

Reviewer 2: ```suggestion
The Insights tool is available for {RHEL} only.
It has no effect on other operating systems.
```

* RHEL 6: `rhel-6-server-rpms`
* RHEL 7: `rhel-7-server-rpms`
* RHEL 8: `rhel-8-for-x86_64-appstream-rpms`
+
The `insights-client` package is installed by default on RHEL 8 except in environments whereby RHEL 8 was deployed with "Minimal Install" option.

. Optional: In the *Token lifetime (hours)* field, change the validity duration of the JSON Web Token (JWT) that {Project} uses for authentication.
The duration of this token defines how long the generated `curl` command works.
You can set the duration to 0 - 999 999 hours or unlimited.
+
Note that {Project} applies the permissions of the user who generates the `curl` command to authorization of hosts.
If the user loses or gains additional permissions, the permissions of the JWT change too.
Therefore, do not delete, block, or change permissions of the user during the token duration.
+
The scope of the JWTs is limited to the registration endpoints only and cannot be used anywhere else.
. Optional: In the *Remote Execution Interface* field, enter the identifier of a network interface that hosts must use for the SSH connection.
If you keep this field blank, {Project} uses the default network interface.

> . Optional: In the *Install packages* field, list the packages (separated with spaces) that you want to install on the host upon registration. This can be set by the `host_packages` parameter.

Reviewer 2: ```suggestion
. Optional: In the *Install packages* field, list the packages (separated with spaces) that you want to install on the host upon registration.
This can be set by the `host_packages` parameter.
```

> . Optional: Select the *Update packages* option to update all packages on the host upon registration. This can be set by the `host_update_packages` parameter.

Reviewer 2: ```suggestion
. Optional: Select the *Update packages* option to update all packages on the host upon registration.
This can be set by the `host_update_packages` parameter.
```

> . Optional: In the *Repository* field, enter a repository to be added before the registration is performed. For example, it can be useful to make the subscription-manager packages available for the purpose of the registration.

Reviewer 2: ```suggestion
. Optional: In the *Repository* field, enter a repository to be added before the registration is performed.
For example, it can be useful to make the `subscription-manager` packages available for the purpose of the registration.
```
Does `subscription-manager` consist of multiple packages?

Author: @lzap I'm gonna need help on this one, because I have no freaking idea yet :) Thank you!

Reviewer 3: Oh it is only a single package.

Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+
It is recommended to avoid insecure calls.
Instead, you can manually copy and install the CA file on each host before registering the host.
+
To do this, find where {Project} stores the CA file by navigating to *Administer* > *Settings* > *Authentication* and locating the value of the *SSL CA file* setting.
+
Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on hosts and enter the following commands:
+
[options="nowrap", subs="+quotes,attributes"]
----
# update-ca-trust enable
# update-ca-trust
----
+
If an attacker, located in the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JSON Web Tokens (JWT).
Therefore, if you have chosen to deploy SSH keys during registration, the attacker will be able to access the host using the SSH key.
+
The following is an example of the `curl` command with the `--insecure` option:
+
[options="nowrap", subs="+quotes,attributes"]
----

> curl -sS --insecure https://{foreman-example-com}/register ...

Reviewer 3: ```suggestion
# curl -sS https://{foreman-example-com}/register ...
```

endif::[]
. Optional: Select the *Insecure* option, if you want to make the first call insecure.
During this first call, hosts download the CA file from {Project}.
Hosts will use this CA file to connect to {Project} with all future calls making them secure.
+
It is recommended to avoid insecure calls.
Instead, you can manually copy and install the CA file on each host before registering the host.
+
To do this, find where {Project} stores the CA file by navigating to *Administer* > *Settings* > *Authentication* and locating the value of the *SSL CA file* setting.
+
Copy the CA file to the `/etc/pki/ca-trust/source/anchors/` directory on hosts and enter the following commands:
+
[options="nowrap", subs="+quotes,attributes"]
----
# update-ca-trust enable
# update-ca-trust
# curl -sS https://{foreman-example-com}/register ...
----
+
If an attacker, located in the network between {Project} and a host, fetches the CA file from the first insecure call, the attacker will be able to access the content of the API calls to and from the registered host and the JSON Web Tokens (JWT).

> Therefore, if you have chosen to deploy SSH keys during registration, the attacker will be able to access the host using the SSH key.

Reviewer 3: I suggest to move "To do this" after this paragraph.

Then break it up into two snippets: one with the update-ca-trust commands (both) and one with the curl command. That should be divided by some filler, something like "To register a host, use the following command".

Reviewer 3: To elaborate this more (I had a meeting sorry), update-ca-trust and curl is one way to do it, the other way is just insecure curl. So we want these two snippets, while I mentioned breaking up the one on line 66-70, that is actually not necessary. Depending on how you like, you might or might not break it I do not insist. It is just we need to have the curl command twice (with and without the insecure option).

I hope it makes sense.

Author: Got it.
However, I think that the "To do this" paragraph would better fit right after "It is recommended", otherwise the flow doesn't make sense IMHO. Check it out in changed files.
